clash-lib/src/Clash/Rewrite/Types.hs

```diff
@@ -162,83 +170,63 @@ normalizeUltra = clashEnv . Lens.to (opt_ultra . envOpts)
 -- generate fresh variables and unique identifiers. In addition, it keeps track
 -- if a transformation/rewrite has been successfully applied.
 newtype RewriteMonad extra a = R
-  { unR :: RewriteEnv -> RewriteState extra -> Any -> (a,RewriteState extra,Any) }
+  { unR :: RWST RewriteEnv Any (RewriteState extra) IO a }
+  deriving newtype
+    ( Applicative
+    , Functor
+    , Monad
+    , MonadFix
+    )
 
 -- | Run the computation in the RewriteMonad
 runR
   :: RewriteMonad extra a
   -> RewriteEnv
   -> RewriteState extra
-  -> (a, RewriteState extra, Any)
-runR m r s = unR m r s mempty
-
-instance MonadFail (RewriteMonad extra) where
-  fail err = error ("RewriteMonad.fail: " ++ err)
-
-instance Functor (RewriteMonad extra) where
-  fmap f m = R $ \ r s w -> case unR m r s w of (a, s', w') -> (f a, s', w')
-  {-# INLINE fmap #-}
-
-instance Applicative (RewriteMonad extra) where
-  pure a = R $ \ _ s w -> (a, s, w)
-  {-# INLINE pure #-}
-  R mf <*> R mx = R $ \ r s w -> case mf r s w of
-    (f,s',w') -> case mx r s' w' of
-      (x,s'',w'') -> (f x, s'', w'')
-  {-# INLINE (<*>) #-}
-
-instance Monad (RewriteMonad extra) where
-  return a = R $ \ _ s w -> (a, s, w)
-  {-# INLINE return #-}
-  m >>= k  =
-    R $ \ r s w -> case unR m r s w of
-      (a,s',w') -> unR (k a) r s' w'
-  {-# INLINE (>>=) #-}
+  -> IO (a, RewriteState extra, Any)
+runR m = RWS.runRWST (unR m)
 
+#if MIN_VERSION_transformers(0,5,6)
```

Comment:
```suggestion
#if !MIN_VERSION_transformers(2,3,0)
```
The CPS instances will be added in the 2.3 release of `mtl`; this future-proofs us when we bump the `mtl` bounds in `clash-lib.cabal`